### PR TITLE
[AC-5486] RoleBin and capacity for queue selection

### PR DESCRIPTION
**Pull request by mc-nathan**
Add RoleBin
Add capacity to queue selection process
Make queue weight management more explicit

### CODE REVIEW

File: classes/allocator.py [hunk 1]
```diff
@@ -52,22 +58,31 @@ def add_startups(self):
...
+        for bin in self.bins:
+            bin.calc_capacity(self.judges)
+
     def work_left(self):
         return any([bin.work_left() for bin in self.bins])
 
     def default_bins(self):
         self.bins = (
             [ReadsBin(),
-             FemaleBin(weight=2*BIN_DEFAULT_WEIGHT),
```
Comment (jonkiparsky) [inline classes/allocator.py:61]: Are we losing the Female Judge criterion here, or is it being put back in some place I'm not seeing?

File: classes/bin.py [hunk 2]
```diff
@@ -28,9 +28,20 @@ def add_startup(self, startup):
...
+    def calc_capacity(self, judges):
+        self.capacity = 0
+        for judge in judges:
+            if self.weight(judge):
+                self.capacity += int(judge.properties["commitment"])
+
+    def match(self, entity):
         return True
 
+    def update(self, judge, startup, keep):
```
Comment (jonkiparsky) [inline classes/bin.py:40]: I wonder if there's a more specific name for what we're doing here. "Update" is pretty generic. Not a blocker.

File: classes/bin.py [hunk 3]
```diff
@@ -28,9 +28,20 @@ def add_startup(self, startup):
...
+        if self.weight(judge):
+            self.update_startup(startup, keep)
+            self.capacity = max(1, self.capacity - 1)
```
Comment (jonkiparsky) [inline classes/bin.py:43]: Do we also need to update other bins that are counting this judge's commitments in their capacity?

File: classes/bin.py [hunk 4]
```diff
@@ -40,10 +51,16 @@ def update_startup(self, startup, keep=False):
     def work_left(self):
         return len(self.queue) > 0
 
-    def weight(self, judge):
-        if self.work_left():
+    def weight(self, judge, field=None):
```
Comment (jonkiparsky) [inline classes/bin.py]: what is the `field` parameter, and why isn't it being used?

File: classes/industry_bin.py [hunk 5]
```diff
@@ -12,9 +12,5 @@ def __init__(self, value, weight=BIN_DEFAULT_WEIGHT):
...
         return "{} Industry Bin".format(self.industry)
 
-    def match(self, startup):
-        return startup.properties["industry"] == self.industry
-
-    def weight(self, judge):
-        if judge.properties["industry"] == self.industry:
-            return super().weight(judge)
+    def match(self, entity):
+        return entity.properties["industry"] == self.industry
```
Comment (jonkiparsky) [inline classes/industry_bin.py:16]: This seems like a reasonable opportunity to get rid of an inline literal string. 
Not a blocker, if you don't do it now, one of us will do it soon.

File: classes/satisfied_bin.py [hunk 6]
```diff
@@ -14,5 +14,8 @@ def update_startup(self, startup, keep=False):
     def weight(self, judge):
         return BIN_LOW_WEIGHT
 
+    def adjusted_weight(self, judge):
+        return self.weight(judge)/float(self.capacity)
```
Comment (jonkiparsky) [inline classes/satisfied_bin.py:18]: Will every bin have an adjusted weight? And will this be at the bottom of that calculation in general?

**Review by jonkiparsky — CHANGES_REQUESTED** [review]
Minor quibbles. The main one is the loss of the Female Judge queue - is this intentional?

File: classes/allocator.py (hunk 1 above)
Comment (mc-nathan) [reply classes/allocator.py:61]: Good catch.  Copy/paste error.

File: classes/bin.py (hunk 2 above)
Comment (mc-nathan) [reply classes/bin.py:40]: I'm about to refactor that, so I'll not change it now.

File: classes/bin.py (hunk 3 above)
Comment (mc-nathan) [reply classes/bin.py:43]: I think that's handled by Judge.finish_startups unless I'm missing something.

File: classes/bin.py (hunk 4 above)
Comment (mc-nathan) [reply classes/bin.py]: Good catch.  Lingering cruft from a refactor.

File: classes/industry_bin.py (hunk 5 above)
Comment (mc-nathan) [reply classes/industry_bin.py:16]: Yeah, I think that's going to part of the refactor to UniversalQueue and SpecificQueue or some such.

File: classes/satisfied_bin.py (hunk 6 above)
Comment (mc-nathan) [reply classes/satisfied_bin.py:18]: SatisfiedBin is a special case.  The one in Bin is the more general answer.

File: classes/bin.py (hunk 3 above)
Comment (jonkiparsky) [reply classes/bin.py:43]: Got it. It was not obvious at first that this is only called by `finish_startups`

**Review by jonkiparsky — APPROVED** [review]

**Pull request merged**